modules/broker-amqp/src/main/java/org/wso2/broker/amqp/codec/handlers/AmqpConnectionHandler.java

```diff
@@ -46,15 +48,20 @@
     private final Map<Integer, AmqpChannel> channels = new HashMap<>();
     private final AmqpServerConfiguration configuration;
     private final Broker broker;
+    private final AmqpMetricManager metricManager;
     private SaslServer saslServer = null;
 
-    public AmqpConnectionHandler(AmqpServerConfiguration configuration, Broker broker) {
+    public AmqpConnectionHandler(AmqpServerConfiguration configuration,
+                                 Broker broker,
+                                 AmqpMetricManager metricManager) {
         this.configuration = configuration;
         this.broker = broker;
+        this.metricManager = metricManager;
+        metricManager.incrementConnectionCount();
     }
 
     public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
-        ctx.channel().closeFuture().addListener(future -> closeConnection(ctx));
+        ctx.channel().closeFuture().addListener(future -> ctx.fireChannelRead((BlockingTask) this::onConnectionClose));
```

Comment:
Does the metric manager count operations involve I/O? Or does it do I/O operations periodically through a separate thread?

Reply:
AFAIK reporting is happening in a separate thread.

Reply:
In that case why do we need to do an async call to `onConnectionClose`?

Reply:
This call propagating to the broker for closing the consumers which involve locks. To avoid blocking the netty event loop I used a blocking task.

Reply:
:+1: